dumbo/src/pdu/udp.rs

```diff
@@ -0,0 +1,328 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Contains support for parsing and writing User Datagram Protocol (UDP) packets
+//!
+//! Details of the UDP packet specification can be found at [1] [2]
+//!
+//! [1]: https://tools.ietf.org/html/rfc768
+//! [2]: https://tools.ietf.org/html/rfc5405
+
+use std::net::Ipv4Addr;
+
+use pdu::bytes::NetworkBytesMut;
+use pdu::ipv4::PROTOCOL_UDP;
+use pdu::Incomplete;
+
+use super::bytes::{InnerBytes, NetworkBytes};
+
+const SOURCE_PORT_OFFSET: usize = 0;
+const DESTINATION_PORT_OFFSET: usize = 2;
+const LENGTH_OFFSET: usize = 4;
+const CHECKSUM_OFFSET: usize = 6;
+const PAYLOAD_OFFSET: usize = 8;
+/// The header length is 8 octets (bytes)
+pub const UDP_HEADER_SIZE: usize = 8;
+
+// A UDP datagram is carried in a single IP packet and is hence limited
+// to a maximum payload of 65,507 bytes for IPv4 and 65,527 bytes for IPv6 [2]
+const IPV4_MAX_UDP_PACKET_SIZE: usize = 65507;
+
+/// Represents errors which may occur while parsing or writing a datagram.
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// Invalid checksum
+    Checksum,
+    /// The specified byte sequence is shorter than the Ethernet header length.
+    DatagramTooShort,
+    /// The payload to be added to the UDP packet exceeds the size allowed
+    /// by the used IP version
+    PayloadTooBig,
+}
+
+/// Interprets the inner bytes as a UDP datagram.
+pub struct UdpDatagram<'a, T: 'a> {
+    bytes: InnerBytes<'a, T>,
+}
+
+impl<'a, T: NetworkBytes> UdpDatagram<'a, T> {
+    /// Interprets `bytes` as a UDP datagram without any validity checks.
+    ///
+    /// # Panics
+    ///
+    ///  This method does not panic, but further method calls on the resulting object may panic if
+    /// `bytes` contains invalid input.
+    #[inline]
+    pub fn from_bytes_unchecked(bytes: T) -> Self {
+        UdpDatagram {
+            bytes: InnerBytes::new(bytes),
+        }
+    }
+
+    /// Interprets `bytes` as a UDP datagram if possible or returns
+    /// the reason for failing to do so
+    #[inline]
+    pub fn from_bytes(
+        bytes: T,
+        verify_checksum: Option<(Ipv4Addr, Ipv4Addr)>,
+    ) -> Result<Self, Error> {
+        if bytes.len() < UDP_HEADER_SIZE {
+            return Err(Error::DatagramTooShort);
+        }
+
+        let datagram = UdpDatagram::from_bytes_unchecked(bytes);
+        if let Some((src_addr, dst_addr)) = verify_checksum {
+            if datagram.compute_checksum(src_addr, dst_addr) != 0 {
+                return Err(Error::Checksum);
+            }
+        }
+
+        // src port (optional): If not used, a value of zero is inserted.
+        // src port be assumed  to be the port  to which a reply should
+        // be addressed  in the absence of any other information (?)
+        Ok(datagram)
+    }
+
+    /// Returns the source port of the UDP datagram
```

Comment:
Would be nice to add a full stop at the end of all these comments that don't have one.